test/integration/daemonset/daemonset_test.go

```diff
@@ -393,3 +397,267 @@ func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
+		nodeClient := clientset.CoreV1().Nodes()
+		podInformer := informers.Core().V1().Pods().Informer()
+		stopCh := make(chan struct{})
+		defer close(stopCh)
+		informers.Start(stopCh)
+		go dc.Run(5, stopCh)
+
+		dsName := "daemonset"
+		ds := newDaemonSet(dsName, ns.Name)
+		ds.Spec.UpdateStrategy = *strategy
+		ds, err := dsClient.Create(ds)
+		if err != nil {
+			t.Fatalf("failed to create daemonset %q: %v", dsName, err)
+		}
+
+		nodeName := "single-node"
+		node := newNode(nodeName, nil)
+		_, err = nodeClient.Create(node)
+		if err != nil {
+			t.Fatalf("failed to create node %q: %v", nodeName, err)
+		}
+
+		// Validate everything works correctly, include marking pods
+		// as ready and ensure they launch on every node
+		validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
+		validateDaemonSetStatus(dsClient, ds.Name, ds.Namespace, 1, t)
+
+		// Set the pod's phase to "Failed"
+		pods := getPods(t, podClient, testLabels())
+		pod := updatePodStatus(t, podClient, pods.Items[0].Name, func(pod *v1.Pod) {
+			pod.Status.Phase = v1.PodFailed
+		})
+
+		// Validate everything works correctly again
+		validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
+		validateDaemonSetStatus(dsClient, ds.Name, ds.Namespace, 1, t)
+
+		// Validate the failed pod is re-created
+		pods = getPods(t, podClient, testLabels())
+		pod = &pods.Items[0]
+		if pod.Status.Phase != v1.PodRunning {
+			t.Fatalf("failed to retry create failed pod %q", pod.Name)
+		}
+	}
+}
+
+// A RollingUpdate DaemonSet should rollback without unnecessary restarts
+func TestRollbackWithoutUnnecessaryRestarts(t *testing.T) {
+	server, closeFn, dc, informers, clientset := setup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("rollback-without-unnecessary-restarts-test", server, t)
+	defer framework.DeleteTestingNamespace(ns, server, t)
+
+	dsClient := clientset.ExtensionsV1beta1().DaemonSets(ns.Name)
+	podClient := clientset.CoreV1().Pods(ns.Name)
+	nodeClient := clientset.CoreV1().Nodes()
+	podInformer := informers.Core().V1().Pods().Informer()
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informers.Start(stopCh)
+	go dc.Run(5, stopCh)
+
+	// Create a rolling update daemonset
+	dsName := "daemonset"
+	ds := newDaemonSet(dsName, ns.Name)
+	ds.Spec.UpdateStrategy = *newRollbackStrategy()
+	oldImage := "image"
+	ds.Spec.Template.Spec.Containers[0].Image = oldImage
+	ds, err := dsClient.Create(ds)
+	if err != nil {
+		t.Fatalf("failed to create daemonset %q: %v", dsName, err)
+	}
+
+	// Create 2 nodes
+	addNodes(nodeClient, 0, 2, nil, t)
+
+	// Validate everything works correctly, include marking pods
+	// as ready and ensure they launch on every node
+	validateDaemonSetPodsAndMarkReady(podClient, podInformer, 2, t)
+	validateDaemonSetStatus(dsClient, ds.Name, ds.Namespace, 2, t)
+
+	// Manually update image of first pod to the new image
+	// Note: daemonset image update does not propagate to its pods
```

Comment:
I spent around 2 days finding a way for daemonset image update to propagate to its pods without manually changing the pods, but failed. I even tried to delete the daemonset, but the daemonset failed to get deleted (you can still get the daemonset after the delete operation succeeds). This problem does not exist for deployment and replicaset, so I speculate that I miss out something important that is unique to daemonset.

As this test is about rolling update daemonset, the pods should be automatically deleted and replaced with new ones when there is a change to daemonset spec. However, coming out with something below does not work:
```
// the image change does not propagate to the pods
newDS := updateDS(t, dsClient, dsName, func(update *v1beta1.DaemonSet){
	update.Spec.Template.Spec.Containers[0].Image = newImage
})

// this checks whether there is at least one pod with newImage, but the check fails (all pods still have old image)
checkAtLeastOnePodWithImage(podClient, newImage, t)
```

The pods are not deleted and replaced even after a long polling timeout (3 minutes).